libs/brig-types/src/Brig/Types/User.hs

```diff
@@ -273,34 +279,98 @@ instance ToJSON SelfProfile where
 
-data RichInfo = RichInfo
-    { richInfoFields :: ![RichField]  -- ^ An ordered list of fields
-    }
-    deriving (Eq, Show, Generic)
+newtype RichInfoAssocList = RichInfoAssocList [RichField]
+  deriving (Eq, Show, Generic)
 
-instance ToJSON RichInfo where
-    toJSON u = object
-        [ "fields" .= richInfoFields u
-        , "version" .= (0 :: Int)
-        ]
+toRichInfoAssocList :: RichInfo -> RichInfoAssocList
+toRichInfoAssocList (RichInfo mp al) = RichInfoAssocList . nubrf $ toal mp <> al
+  where
+    nubrf :: [RichField] -> [RichField]
+    nubrf = nubBy ((==) `on` \(RichField k _) -> CI.mk k)
+
+    toal :: Map (CI Text) Text -> [RichField]
+    toal = map (\(k, v) -> RichField (CI.foldedCase k) v) . Map.toAscList
+
+richInfoMapURN, richInfoAssocListURN :: Text
+richInfoMapURN = "urn:ietf:params:scim:schemas:extension:wire:1.0:User"
+richInfoAssocListURN = "urn:wire:scim:schemas:profile:1.0"
 
 instance FromJSON RichInfo where
-    parseJSON = withObject "RichInfo" $ \o -> do
-        version :: Int <- o .: "version"
-        case version of
-            0 -> do
-                fields <- o .: "fields"
-                checkDuplicates (map richFieldType fields)
-                pure (RichInfo fields)
-            _ -> fail ("unknown version: " <> show version)
-      where
-        checkDuplicates :: [Text] -> Aeson.Parser ()
-        checkDuplicates xs =
-            case filter ((> 1) . length) . group . sort $ xs of
-                [] -> pure ()
-                ds -> fail ("duplicate fields: " <> show (map head ds))
+  parseJSON =
+    withObject "RichInfo"
+    $ \o ->
+        let objWithCIKeys = hmMapKeys CI.mk o
+        in normalizeRichInfo
+           <$> (RichInfo
+                <$> extractMap objWithCIKeys
+                <*> extractAssocList objWithCIKeys)
+    where
+      extractMap :: HashMap (CI Text) Value -> Aeson.Parser (Map (CI Text) Text)
+      extractMap o =
+        case HM.lookup (CI.mk richInfoMapURN) o of
+          Nothing -> pure mempty
+          Just innerObj -> do
+            Map.mapKeys CI.mk <$> parseJSON innerObj
+
+      extractAssocList :: HashMap (CI Text) Value -> Aeson.Parser [RichField]
+      extractAssocList o =
+        case HM.lookup (CI.mk richInfoAssocListURN) o of
+          Nothing -> pure []
+          Just (Object innerObj) -> do
+            richInfo <- mLookup "richinfo" $ hmMapKeys CI.mk innerObj
+            case richInfo of
+              Object richinfoObj -> do
+                fields <- richInfoAssocListFromObject richinfoObj
+                pure fields
+              Array fields -> parseJSON (Array fields)
+              v -> Aeson.typeMismatch "Object" v
+          Just v -> Aeson.typeMismatch "Object" v
+
+      hmMapKeys :: (Eq k2, Hashable k2) => (k1 -> k2) -> HashMap k1 v -> HashMap k2 v
+      hmMapKeys f = HashMap.fromList . (map (\(k,v) -> (f k, v))) . HashMap.toList
+
+      mLookup :: (MonadFail m, Show k, Eq k, Hashable k) => k -> HashMap k v -> m v
```

Comment:
```suggestion
      hmLookup :: (MonadFail m, Show k, Eq k, Hashable k) => k -> HashMap k v -> m v
```

(for consistency)

Reply:
By `mLookup` I meant monadic lookup, `hmLookup` is too close to `HM.lookup` which is lookup that returns a `Maybe`. I am going to rename it to `lookupOrFail`.